### PR TITLE
Fix clippy warnings

### PR DESCRIPTION
**Pull request by yutannihilation**
Now it's a good time to look into what warnings clippy complains; some of them should be fixed, and the other might be ignorable.

full warnings: https://gist.github.com/yutannihilation/acc2fb7f4b18d2aba79cf86ae89e4654#file-clippy_warnings_20210320-txt-L360

I still see these warnings, which I don't know how to fix:

```
warning: unsafe function's docs miss `# Safety` section
  --> extendr-api/src/ownership.rs:24:1
   |
24 | / pub unsafe fn protect(sexp: SEXP) {
25 | |     let mut own = OWNERSHIP.lock().expect("protect failed");
26 | |     own.protect(sexp);
27 | | }
   | |_^
   |
   = note: `#[warn(clippy::missing_safety_doc)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_safety_doc

warning: unsafe function's docs miss `# Safety` section
  --> extendr-api/src/ownership.rs:29:1
   |
29 | / pub unsafe fn unprotect(sexp: SEXP) {
30 | |     let mut own = OWNERSHIP.lock().expect("unprotect failed");
31 | |     own.unprotect(sexp);
32 | | }
   | |_^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_safety_doc

warning: item `robj::Robj` has a public `len` method but no corresponding `is_empty` method
   --> extendr-api/src/robj/mod.rs:146:1
    |
146 | / impl Robj {
147 | |     /// Get a copy of the underlying SEXP.
148 | |     /// Note: this is unsafe.
149 | |     #[doc(hidden)]
...   |
689 | |     }
690 | | }
    | |_^
    |
    = note: `#[warn(clippy::len_without_is_empty)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#len_without_is_empty

warning: methods called `to_*` usually take self by reference; consider choosing a less ambiguous name
   --> extendr-api/src/robj/mod.rs:684:21
    |
684 |     pub fn to_owned(self) -> Robj {
    |                     ^^^^
    |
    = note: `#[warn(clippy::wrong_self_convention)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention

warning: method `from_str` can be confused for the standard trait method `std::str::FromStr::from_str`
  --> extendr-api/src/wrapper/character.rs:22:5
   |
22 | /     pub fn from_str(val: &str) -> Self {
23 | |         unsafe {
24 | |             Character {
25 | |                 robj: new_owned(str_to_character(val)),
26 | |             }
27 | |         }
28 | |     }
   | |_____^
   |
   = note: `#[warn(clippy::should_implement_trait)]` on by default
   = help: consider implementing the trait `std::str::FromStr` or choosing a less ambiguous method name
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#should_implement_trait

warning: methods called `new` usually take no self; consider choosing a less ambiguous name
  --> extendr-api/src/wrapper/environment.rs:17:16
   |
17 |     pub fn new(parent: Environment) -> Self {
   |                ^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention

warning: methods called `from_*` usually take no self; consider choosing a less ambiguous name
  --> extendr-api/src/wrapper/environment.rs:58:27
   |
58 |     pub fn from_pairs<NV>(parent: Environment, names_and_values: NV) -> Self
   |                           ^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention

warning: method `from_str` can be confused for the standard trait method `std::str::FromStr::from_str`
  --> extendr-api/src/wrapper/primitive.rs:30:5
   |
30 | /     pub fn from_str(val: &str) -> Result<Self> {
31 | |         single_threaded(|| unsafe {
32 | |             // Primitives have a special "SYMVALUE" entry in their symbol.
33 | |             let sym = Symbol::from_str(val);
...  |
40 | |         })
41 | |     }
   | |_____^
   |
   = help: consider implementing the trait `std::str::FromStr` or choosing a less ambiguous method name
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#should_implement_trait

warning: method `from_str` can be confused for the standard trait method `std::str::FromStr::from_str`
  --> extendr-api/src/wrapper/symbol.rs:32:5
   |
32 | /     pub fn from_str(val: &str) -> Self {
33 | |         if val.is_empty() || val.is_na() {
34 | |             unbound_value()
35 | |         } else {
...  |
39 | |         }
40 | |     }
   | |_____^
   |
   = help: consider implementing the trait `std::str::FromStr` or choosing a less ambiguous method name
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#should_implement_trait

warning: 9 warnings emitted

    Finished dev [unoptimized + debuginfo] target(s) in 0.03s
```

### CODE REVIEW

**Review by clauswilke — APPROVED** [review]
Looks generally good to me, but somebody else should review also.

**Comment by clauswilke** [comment]
The lints about the missing `# Safety` section are about the documentation. Every exported unsafe function should have a section where the safety implications of the function are explained. See here: https://github.com/rust-lang/rust-clippy/blob/0bdaa77d958d91a6d1328b438ca4e15758fbc6ed/clippy_lints/src/doc.rs#L68-L95

**Comment by yutannihilation** [comment]
Thanks. For safety section, I think I have no enough knowledge to write it and probably it will be a important topic to show the users what to care in terms of safety. Let me file a new issue for it.

I added some more minor fixes for warnings found by rerunning clippy on nightly.

**Comment by yutannihilation** [comment]
@andy-thomason 
Could you take a look when you have time? I think these changes are safe to merge, but, as it's a matter of preferences, you might have some opinion about the coding styles.

**Comment by yutannihilation** [comment]
Filed separate issues that I probably cannot solve within this pull request.

* #197: Safety section
* #198: `FromStr` trait
* #199: `Environment::new()` etc.

**Comment by yutannihilation** [comment]
I'll merge this after some more days as the changes are minor ones.

**Review by andy-thomason — APPROVED** [review]
A lot of changes :)

I hope these were largely automatic.

Nothing seems to have changed the function of the code.

I have spotted a couple of potential errors that I should plug
such as the direct use of `Robj::Owned` which now conflicts with the
ownership mechanism.

**Comment by andy-thomason** [comment]
I'll add it to my git pre-commit script.

I already have `cargo fmt`

This will ensure that everything is done prior to pushing.

```
$ cat .git/hooks/pre-commit
#!/bin/sh

cargo fmt
cargo clippy
```

**Pull request merged**
